File: Sorting/InsertionSort/InsertionSort.cpp
#include <iostream>
using namespace std;


/* 	
	Time Complexity Analysis:
	Best Case: O(n) : When the array is already sorted
	Worst Case: O(n^2): When the array is reversed sorted
*/
	
void insertion_sort (int arr[], int n) {

	for (int i = 1; i < n; i++) {

		int key = arr[i];

		int j = i - 1;

		while(j >= 0 && arr[j] > key) {
			arr[j+1] = arr[j];
			j--;
		}
		arr[j+1] = key;
	} 
}

int main()
{

	int arr[] = { 4, 1, 3, 9, 7};

	int n = sizeof(arr)/sizeof(int);

	cout << "Before Sorting: \n";

	for (int i = 0; i < n; i++) 
		cout << arr[i] << " ";	

	cout << "\n";

	insertion_sort(arr, n);

	cout << "After Sorting: \n";
	for (int i = 0; i < n; i++) 
		cout << arr[i] << " ";	
	return 0;
}

File: Sorting/InsertionSort/InsertionSort.java
package Sorting.InsertionSort;

class InsertionSort {
    public static void main(String[] args) {
        
        int [] arr = {5, 4 , 3, 2, 1};

        System.out.println("Before Sorting:");
        for (int i : arr)
            System.out.print(i + " ");

        System.out.println();

        insert_sort(arr);   // Perform Sorting

        System.out.println("After Sorting:");
        for (int i : arr)
            System.out.print(i + " ");

        System.out.println();
        
    }

    private static void insert_sort (int [] arr) {
        for (int i = 1; i < arr.length; i++) {
            int key = arr[i];
            int j = i -1;

            while (j >= 0 && arr[j] > key) {
                arr[j+1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
    }

}